web-api/src/storage/property_filter.rs

```diff
@@ -0,0 +1,230 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! This module contains generic code used by the various aspects of property filtering.
+
+use std::collections::HashMap;
+
+use derive_more::{Display, From, IntoIterator};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::models::DocumentPropertyId;
+
+#[derive(Debug, Display, PartialEq, Error, Serialize)]
+#[non_exhaustive]
+
+pub(crate) enum IncompatibleUpdate {
+    #[display(fmt = "Property {property} is already defined.")]
+    PropertyIsAlreadyIndexed { property: DocumentPropertyId },
+    #[display(
+        fmt = "Only {allowed} indexed properties including publication_date are allowed, got: {count}"
+    )]
+    TooManyProperties { count: usize, allowed: usize },
+}
+
+//Hint: Currently the API and internal definition match so we use the same type.
+#[derive(Debug, Clone, Default, IntoIterator, Serialize, Deserialize)]
+#[serde(transparent)]
+pub(crate) struct IndexedPropertiesSchemaUpdate(
+    #[into_iterator(owned, ref)] IndexedPropertiesSchema,
+);
+
+impl IndexedPropertiesSchemaUpdate {
+    pub(crate) fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+//Hint: Currently the API and internal definition match so we use the same type.
+#[derive(Debug, Clone, Default, IntoIterator, From, Serialize, Deserialize)]
+pub(crate) struct IndexedPropertiesSchema {
+    #[into_iterator(owned, ref)]
+    properties: HashMap<DocumentPropertyId, IndexedPropertyDefinition>,
+}
+
+impl IndexedPropertiesSchema {
+    pub(crate) fn len(&self) -> usize {
+        self.properties.len()
+    }
+
+    pub(crate) fn update(
+        &mut self,
+        schema_update: &IndexedPropertiesSchemaUpdate,
+        max_total_properties: usize,
+    ) -> Result<(), IncompatibleUpdate> {
+        for (property, _) in schema_update {
+            if self.properties.contains_key(property) {
+                return Err(IncompatibleUpdate::PropertyIsAlreadyIndexed {
+                    property: property.clone(),
+                });
+            }
+        }
+        let count = self.len() + schema_update.len();
+        if count > max_total_properties {
+            return Err(IncompatibleUpdate::TooManyProperties {
+                count,
+                allowed: max_total_properties,
+            });
+        }
+        self.properties.extend(schema_update.clone().into_iter());
+        Ok(())
+    }
+}
+
+//Hint: Currently the API and internal definition match so we use the same type.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub(crate) struct IndexedPropertyDefinition {
+    pub(crate) r#type: IndexedPropertyType,
+}
+
+//Hint: Currently the API and internal definition match so we use the same type.
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, sqlx::Type)]
+#[serde(rename_all = "snake_case")]
+#[sqlx(type_name = "indexed_property_type")]
+#[sqlx(rename_all = "snake_case")]
```

Comment:
:art: we could group all sqlx attributes on the type
```suggestion
#[sqlx(rename_all = "snake_case", type_name = "indexed_property_type")]
```